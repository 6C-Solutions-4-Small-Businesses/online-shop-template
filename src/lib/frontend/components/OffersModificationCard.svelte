<script lang="ts">
    import {getAbbreviatedUnit} from '$lib/frontend/core/Helper'
    import type {OffersModificationPresentation} from '$lib/frontend/presentations/OffersModificationPresentation'
    import Image from '$lib/frontend/components/Image.svelte'

    const typeOfModificationToOffers = () => {
        if (offer.deleted) {
            return 'suppression'
        } else if (offer.id) {
            return 'modification'
        } else {
            return 'addition'
        }
    }
    const modificationTypeLabel = () => {
        switch (typeOfModificationToOffers()) {
            case 'suppression':
                return 'supprimé'
            case 'modification':
                return 'modifié'
            case 'addition':
                return 'ajouté'
        }
    }

    const modificationTypeLabelColor = () => {
        switch (typeOfModificationToOffers()) {
            case 'suppression':
                return '#ef4444'
            case 'modification':
                return '#3d83f7'
            case 'addition':
                return '#09cc4c'
        }
    }

    export let offer: OffersModificationPresentation

    function getImagePreviewId() {
        return `${offer.name}-offer-preview-image`
    }

    function offerAbbreviatedUnit() {
        return getAbbreviatedUnit(offer.unit)
    }

    function getPriceFirstPart(price: number): string {
        const pricePart = price.toString().split('.')[0]

        return pricePart ? pricePart : '00'
    }

    function getPriceSecondPart(price: number): string {
        const pricePart = price.toFixed(2).toString().split('.')[1]

        return pricePart ? pricePart : '00'
    }
</script>

<div
        class="w-full bg-white shadow-2xl"
        role="none"
>
    <div class="relative">
        <div class="absolute flex w-24 h-24 top-0 text-white items-center visible z-10">
            <svg height="100" viewBox="0 0 490 490" width="100" x="0px" xml:space="preserve" y="0px">
                  <polygon points="0,0 490,0 0,490" style="{`fill: ${modificationTypeLabelColor()}`}"/>
                </svg>
            <div class="absolute -rotate-45 origin-center top-6 right-4 w-full text-[1rem] text-center font-semibold uppercase">
                {modificationTypeLabel()}
            </div>
        </div>
        {#if offer.imageData}
            <img
                id="{getImagePreviewId()}"
                data-testid="{getImagePreviewId()}"
                class="w-full aspect-1"
                alt="{offer.name}"
                src="{offer.imageData}">
        {:else if offer.image}
            <Image
                    imageRemoteUrl="{offer.image}"
                    name="{offer.name ?? ''}"
                    classes="w-full aspect-1 object-contain"/>
        {/if}
    </div>
    <div class="bottom-0 right-0 w-full flex justify-end">
        <div
                class="filter backdrop-blur-sm bg-green-300/40 font-sans w-full min-h-[4.2rem] justify-between flex items-center text-lg"
                role="region"
        >
            <div class="w-8/12 font-light text-xl px-2 text-wrap overflow-hidden text-overflow: ellipsis;">
                {offer.name}
            </div>
            <div class="w-4/12 flex flex-row text-xl">
                <span class="block">${getPriceFirstPart(offer.price)}
                    <sup>{getPriceSecondPart(offer.price)}</sup>
                </span>
                {#if offerAbbreviatedUnit()}
                    <span class="text-m block text-right pr-1"><sub>/ {offerAbbreviatedUnit()}</sub></span>
                {/if}
            </div>
        </div>
    </div>
</div>
