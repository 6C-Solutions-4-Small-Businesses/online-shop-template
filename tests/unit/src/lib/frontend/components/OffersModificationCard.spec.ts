import {cleanup, render, type RenderResult} from '@testing-library/svelte'
import {afterEach, describe, expect} from 'vitest'
import OffersModificationCard from '$lib/frontend/components/OffersModificationCard.svelte'
import {mock, type MockProxy} from 'vitest-mock-extended'
import type {OffersModificationPresentation} from '$lib/frontend/presentations/OffersModificationPresentation'
import '@testing-library/jest-dom'

describe('OffersModificationCard', () => {

    const offerId = 'ordered-item-id'
    const price = 10099
    const name = 'ordered-item-name'
    const unit = 'Centiliter'

    let offer: MockProxy<OffersModificationPresentation>
    let view: RenderResult<OffersModificationCard, typeof import('@testing-library/dom/types/queries')>

    beforeEach(() => {
        offer = mock<OffersModificationPresentation>({
            id: offerId,
            price,
            name,
            unit,
        })
        view = render(OffersModificationCard, {
            props: getComponentProperties(),
        })
    })

    describe('Structure', () => {

        it('should have match snapshot', () => {

            expect(view.container).toMatchSnapshot()
        })

        describe('when offer image is a URL', () => {

            const imageUrl = 'https://cdn-cemal.nitrocdn.com/iwTMUnSsYeigzDpMWgwdBZhpcIeWwszq/assets/images/optimized/rev-59c8081/www.aaronfaber.com/wp-content/uploads/2017/03/product-placeholder-wp-95907_800x675.jpg'

            beforeEach(() => {
                offer.image = imageUrl
                view.rerender({
                    props: getComponentProperties(),
                })
            })

            it('should show the image', () => {

                const image = view.getByAltText(name) as HTMLImageElement
                expect(image).toBeInTheDocument()
            })
        })

        describe('when offer image is base64 data', () => {

            const imageData = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAABjElEQVRIS+2UwQ3CMAxFX4'

            beforeEach(() => {
                offer.imageData = imageData
                view.rerender({
                    props: getComponentProperties(),
                })
            })

            it('should show the image as well', () => {

                const image = view.getByAltText(name) as HTMLImageElement
                expect(image.src).toStrictEqual(imageData)
            })
        })

        describe('when offer modification has no id and "deleted" undefined', () => {

            beforeEach(() => {
                offer.id = undefined
                offer.deleted = undefined
                view.rerender({
                    props: getComponentProperties(),
                })
            })

            it('should show a banner labeled "ajouté"', () => {

                const banner = view.getByText('ajouté')
                expect(banner).toBeInTheDocument()
            })

            it('should match snapshot', () => {
                expect(view.container).toMatchSnapshot()
            })
        })

        describe('when offer modification has "deleted" property set', () => {

            beforeEach(() => {
                offer.deleted = true
                view.rerender({
                    props: getComponentProperties(),
                })
            })

            it('should show a banner labeled "supprimé"', () => {

                const banner = view.getByText('supprimé')
                expect(banner).toBeInTheDocument()
            })

            it('should match snapshot', () => {
                expect(view.container).toMatchSnapshot()
            })
        })

        describe('when offer modification has an id but not the "deleted" property set', () => {

            beforeEach(() => {
                offer.deleted = undefined
                view.rerender({
                    props: getComponentProperties(),
                })
            })

            it('should show a banner labeled "modifié"', () => {

                const banner = view.getByText('modifié')
                expect(banner).toBeInTheDocument()
            })

            it('should match snapshot', () => {
                expect(view.container).toMatchSnapshot()
            })
        })
    })


    function getComponentProperties() {
        return {
            offer,
        }
    }

    afterEach(() => {
        cleanup()
        vi.clearAllMocks()
    })
})
