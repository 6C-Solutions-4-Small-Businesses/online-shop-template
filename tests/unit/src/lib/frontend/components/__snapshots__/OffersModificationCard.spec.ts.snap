// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OffersModificationCard > Structure > should have match snapshot 1`] = `
<body>
  <div>
    <div
      class="w-full bg-white shadow-2xl"
      role="none"
    >
      <div
        class="relative"
      >
        <div
          class="absolute flex w-24 h-24 top-0 text-white items-center visible z-10"
        >
          <svg
            height="100"
            viewBox="0 0 490 490"
            width="100"
            x="0px"
            xml:space="preserve"
            y="0px"
          >
            <polygon
              points="0,0 490,0 0,490"
              style="fill: #ef4444"
            />
          </svg>
           
          <div
            class="absolute -rotate-45 origin-center top-6 right-4 w-full text-[1rem] text-center font-semibold uppercase"
          >
            supprimé
          </div>
        </div>
         
        <img
          alt="ordered-item-name"
          class="w-full aspect-1"
          data-testid="ordered-item-name-offer-preview-image"
          id="ordered-item-name-offer-preview-image"
          src="function(...s) {
    let r = A(t);
    r.called = !0, r.callCount++, r.calls.push(s);
    let m = r.next.shift();
    if (m) {
      r.results.push(m);
      let [l, o] = m;
      if (l === "ok")
        return o;
      throw o;
    }
    let p, d = "ok";
    if (r.impl)
      try {
        new.target ? p = Reflect.construct(r.impl, s, new.target) : p = r.impl.apply(this, s), d = "ok";
      } catch (l) {
        throw p = l, d = "error", r.results.push([d, l]), l;
      }
    let a = [d, p];
    if (b(p)) {
      let l = p.then((o) => a[1] = o).catch((o) => {
        throw a[0] = "error", a[1] = o, o;
      });
      Object.assign(l, p), p = l;
    }
    return r.results.push(a), p;
  }"
        />
      </div>
       
      <div
        class="bottom-0 right-0 w-full flex justify-end"
      >
        <div
          class="filter backdrop-blur-sm bg-green-300/40 font-sans w-full min-h-[4.2rem] justify-between flex items-center text-lg"
          role="region"
        >
          <div
            class="w-8/12 font-light text-xl px-2 text-wrap overflow-hidden text-overflow: ellipsis;"
          >
            ordered-item-name
          </div>
           
          <div
            class="w-4/12 flex flex-row text-xl"
          >
            <span
              class="block"
            >
              $
              10099
               
              <sup>
                00
              </sup>
            </span>
             
            <span
              class="text-m block text-right pr-1"
            >
              <sub>
                / 
                cl
              </sub>
            </span>
          </div>
        </div>
      </div>
    </div>
    <!--&lt;OffersModificationCard&gt;-->
  </div>
</body>
`;

exports[`OffersModificationCard > Structure > when offer modification has "deleted" property set > should match snapshot 1`] = `
<body>
  <div>
    <div
      class="w-full bg-white shadow-2xl"
      role="none"
    >
      <div
        class="relative"
      >
        <div
          class="absolute flex w-24 h-24 top-0 text-white items-center visible z-10"
        >
          <svg
            height="100"
            viewBox="0 0 490 490"
            width="100"
            x="0px"
            xml:space="preserve"
            y="0px"
          >
            <polygon
              points="0,0 490,0 0,490"
              style="fill: #ef4444"
            />
          </svg>
           
          <div
            class="absolute -rotate-45 origin-center top-6 right-4 w-full text-[1rem] text-center font-semibold uppercase"
          >
            supprimé
          </div>
        </div>
         
        <img
          alt="ordered-item-name"
          class="w-full aspect-1"
          data-testid="ordered-item-name-offer-preview-image"
          id="ordered-item-name-offer-preview-image"
          src="function(...s) {
    let r = A(t);
    r.called = !0, r.callCount++, r.calls.push(s);
    let m = r.next.shift();
    if (m) {
      r.results.push(m);
      let [l, o] = m;
      if (l === "ok")
        return o;
      throw o;
    }
    let p, d = "ok";
    if (r.impl)
      try {
        new.target ? p = Reflect.construct(r.impl, s, new.target) : p = r.impl.apply(this, s), d = "ok";
      } catch (l) {
        throw p = l, d = "error", r.results.push([d, l]), l;
      }
    let a = [d, p];
    if (b(p)) {
      let l = p.then((o) => a[1] = o).catch((o) => {
        throw a[0] = "error", a[1] = o, o;
      });
      Object.assign(l, p), p = l;
    }
    return r.results.push(a), p;
  }"
        />
      </div>
       
      <div
        class="bottom-0 right-0 w-full flex justify-end"
      >
        <div
          class="filter backdrop-blur-sm bg-green-300/40 font-sans w-full min-h-[4.2rem] justify-between flex items-center text-lg"
          role="region"
        >
          <div
            class="w-8/12 font-light text-xl px-2 text-wrap overflow-hidden text-overflow: ellipsis;"
          >
            ordered-item-name
          </div>
           
          <div
            class="w-4/12 flex flex-row text-xl"
          >
            <span
              class="block"
            >
              $
              10099
               
              <sup>
                00
              </sup>
            </span>
             
            <span
              class="text-m block text-right pr-1"
            >
              <sub>
                / 
                cl
              </sub>
            </span>
          </div>
        </div>
      </div>
    </div>
    <!--&lt;OffersModificationCard&gt;-->
  </div>
</body>
`;

exports[`OffersModificationCard > Structure > when offer modification has an id but not the "deleted" property set > should match snapshot 1`] = `
<body>
  <div>
    <div
      class="w-full bg-white shadow-2xl"
      role="none"
    >
      <div
        class="relative"
      >
        <div
          class="absolute flex w-24 h-24 top-0 text-white items-center visible z-10"
        >
          <svg
            height="100"
            viewBox="0 0 490 490"
            width="100"
            x="0px"
            xml:space="preserve"
            y="0px"
          >
            <polygon
              points="0,0 490,0 0,490"
              style="fill: #3d83f7"
            />
          </svg>
           
          <div
            class="absolute -rotate-45 origin-center top-6 right-4 w-full text-[1rem] text-center font-semibold uppercase"
          >
            modifié
          </div>
        </div>
         
        <img
          alt="ordered-item-name"
          class="w-full aspect-1"
          data-testid="ordered-item-name-offer-preview-image"
          id="ordered-item-name-offer-preview-image"
          src="function(...s) {
    let r = A(t);
    r.called = !0, r.callCount++, r.calls.push(s);
    let m = r.next.shift();
    if (m) {
      r.results.push(m);
      let [l, o] = m;
      if (l === "ok")
        return o;
      throw o;
    }
    let p, d = "ok";
    if (r.impl)
      try {
        new.target ? p = Reflect.construct(r.impl, s, new.target) : p = r.impl.apply(this, s), d = "ok";
      } catch (l) {
        throw p = l, d = "error", r.results.push([d, l]), l;
      }
    let a = [d, p];
    if (b(p)) {
      let l = p.then((o) => a[1] = o).catch((o) => {
        throw a[0] = "error", a[1] = o, o;
      });
      Object.assign(l, p), p = l;
    }
    return r.results.push(a), p;
  }"
        />
      </div>
       
      <div
        class="bottom-0 right-0 w-full flex justify-end"
      >
        <div
          class="filter backdrop-blur-sm bg-green-300/40 font-sans w-full min-h-[4.2rem] justify-between flex items-center text-lg"
          role="region"
        >
          <div
            class="w-8/12 font-light text-xl px-2 text-wrap overflow-hidden text-overflow: ellipsis;"
          >
            ordered-item-name
          </div>
           
          <div
            class="w-4/12 flex flex-row text-xl"
          >
            <span
              class="block"
            >
              $
              10099
               
              <sup>
                00
              </sup>
            </span>
             
            <span
              class="text-m block text-right pr-1"
            >
              <sub>
                / 
                cl
              </sub>
            </span>
          </div>
        </div>
      </div>
    </div>
    <!--&lt;OffersModificationCard&gt;-->
  </div>
</body>
`;

exports[`OffersModificationCard > Structure > when offer modification has no id and "deleted" undefined > should match snapshot 1`] = `
<body>
  <div>
    <div
      class="w-full bg-white shadow-2xl"
      role="none"
    >
      <div
        class="relative"
      >
        <div
          class="absolute flex w-24 h-24 top-0 text-white items-center visible z-10"
        >
          <svg
            height="100"
            viewBox="0 0 490 490"
            width="100"
            x="0px"
            xml:space="preserve"
            y="0px"
          >
            <polygon
              points="0,0 490,0 0,490"
              style="fill: #09cc4c"
            />
          </svg>
           
          <div
            class="absolute -rotate-45 origin-center top-6 right-4 w-full text-[1rem] text-center font-semibold uppercase"
          >
            ajouté
          </div>
        </div>
         
        <img
          alt="ordered-item-name"
          class="w-full aspect-1"
          data-testid="ordered-item-name-offer-preview-image"
          id="ordered-item-name-offer-preview-image"
          src="function(...s) {
    let r = A(t);
    r.called = !0, r.callCount++, r.calls.push(s);
    let m = r.next.shift();
    if (m) {
      r.results.push(m);
      let [l, o] = m;
      if (l === "ok")
        return o;
      throw o;
    }
    let p, d = "ok";
    if (r.impl)
      try {
        new.target ? p = Reflect.construct(r.impl, s, new.target) : p = r.impl.apply(this, s), d = "ok";
      } catch (l) {
        throw p = l, d = "error", r.results.push([d, l]), l;
      }
    let a = [d, p];
    if (b(p)) {
      let l = p.then((o) => a[1] = o).catch((o) => {
        throw a[0] = "error", a[1] = o, o;
      });
      Object.assign(l, p), p = l;
    }
    return r.results.push(a), p;
  }"
        />
      </div>
       
      <div
        class="bottom-0 right-0 w-full flex justify-end"
      >
        <div
          class="filter backdrop-blur-sm bg-green-300/40 font-sans w-full min-h-[4.2rem] justify-between flex items-center text-lg"
          role="region"
        >
          <div
            class="w-8/12 font-light text-xl px-2 text-wrap overflow-hidden text-overflow: ellipsis;"
          >
            ordered-item-name
          </div>
           
          <div
            class="w-4/12 flex flex-row text-xl"
          >
            <span
              class="block"
            >
              $
              10099
               
              <sup>
                00
              </sup>
            </span>
             
            <span
              class="text-m block text-right pr-1"
            >
              <sub>
                / 
                cl
              </sub>
            </span>
          </div>
        </div>
      </div>
    </div>
    <!--&lt;OffersModificationCard&gt;-->
  </div>
</body>
`;
